### PR TITLE
Optimize the code to remove closure and bind

### PR DESCRIPTION
**Pull request by raymondfeng**
/to @bajtos or @ritch 
/cc @bnoordhuis

### CODE REVIEW

**Comment by ritch** [comment]
Did this result in any serious performance improvements?

The changes seem fine to me.

**Comment by raymondfeng** [comment]
I tested the changes with JMeter on 16 threads. Throughput increased from 2383/s to 2576/s.

File: lib/http-context.js [hunk 1]
```diff
@@ -192,27 +191,45 @@ function coerce(str) {
...
+  var i;
+  var n;
 
   switch (type) {
     case 'string':
       return coerce(obj);
     case 'object':
       if (obj) {
-        Object.keys(obj).forEach(function(key) {
+        var props = Object.keys(obj);
```
Comment (bnoordhuis) [inline lib/http-context.js]: If obj is prototype-less or if you don't care that prototype properties get iterated over, using `for (var key in obj)` is usually faster.

File: lib/http-context.js [hunk 2]
```diff
@@ -192,27 +191,45 @@ function coerce(str) {
...
+        for (i = 0, n = props.length; i < n; i++) {
```
Comment (bnoordhuis) [inline lib/http-context.js]: I don't know what the house style is but just you can just use `var i` here and below.  The declaration gets hoisted to the top and it doesn't matter from a performance perspective.

File: lib/http-context.js [hunk 3]
```diff
@@ -192,27 +191,45 @@ function coerce(str) {
...
+        coerceAll(obj[i]);
+      }
       break;
   }
 
   return obj;
 }
 
+function buildArgs(ctx, method, fn) {
+  var args = ctx.args || {};
```
Comment (bnoordhuis) [inline lib/http-context.js]: This assignment looks dead to me.  It's reassigned in the try block and it's unused in the catch block.

Comment (raymondfeng) [inline lib/http-context.js]: Good catch.

File: lib/http-context.js (hunk 2 above)
Comment (raymondfeng) [inline lib/http-context.js]: We declare the variables at the top.

File: lib/http-context.js (hunk 1 above)
Comment (raymondfeng) [inline lib/http-context.js]: I was following https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#5-for-in

Comment (bnoordhuis) [inline lib/http-context.js]: Petka is not wrong but it's a tradeoff.  Garbage collection is frequently the biggest performance killer, hence my advice.  Try tracing it with --trace_opt and --trace_deopt.  If it does deoptimize, move it into a separate function.

**Pull request merged**